utils/rarity.js

```diff
@@ -1,82 +1,114 @@
-  let layers = config.layersOrder;
+  // loop through the traits
+  for (let j = 0; j < traits.length; j++) {
+    // get the current trait
+    const trait = traits[j];
+    // get the trait type
+    const trait_type = trait.trait_type;
+    // get the value of each trait_type
+    const value = trait.value;
+    // get the count of each trait_type and value
+    const trait_type_value_count = metadata.filter((nft) =>
+      nft.attributes.some(
+        (trait) => trait.trait_type === trait_type && trait.value === value
+      )
+    ).length;
+    // calculate the score of each trait_type and value aka rarity score read more at: https://raritytools.medium.com/ranking-rarity-understanding-rarity-calculation-methods-86ceaeb9b98c
+    const trait_individual_score = 1 / (trait_type_value_count / metadata.length);
+    // round the score to two decimal places
+    const trait_individual_score_final = Math.round(trait_individual_score * 100) / 100;
 
-  layers.forEach((layer) => {
-    // get elements for each layer
-    let elementsForLayer = [];
-    let elements = getElements(`${layersDir}/${layer.name}/`);
-    elements.forEach((element) => {
-      // just get name and weight for each element
-      let rarityDataElement = {
-        trait: element.name,
-        weight: element.weight.toFixed(0),
-        occurrence: 0, // initialize at 0
-      };
-      elementsForLayer.push(rarityDataElement);
+    // push the data to the data array
+    data.push({
+      trait_type: trait_type,
+      value: value,
+      count: trait_type_value_count,
+      score: trait_individual_score_final,
     });
-    let layerName =
-      layer.options?.["displayName"] != undefined
-        ? layer.options?.["displayName"]
-        : layer.name;
-    // don't include duplicate layers
-    if (!rarityData.includes(layer.name)) {
-      // add elements for each layer to chart
-      rarityData[layerName] = elementsForLayer;
-    }
-  });
-});
+  }
 
-// fill up rarity chart with occurrences from metadata
-data.forEach((element) => {
-  let attributes = element.attributes;
-  attributes.forEach((attribute) => {
-    let traitType = attribute.trait_type;
-    let value = attribute.value;
+  // if the ranked is true then sort the data by total_score and add ranking
+  if (rarityConfigurations.ranked) {
+    output.sort((a, b) => b.total_score - a.total_score);
+    for (let i = 0; i <= output.length - 1; i++) {
+      output[i].rank = i + 1;
+    }
+  }
 
-    let rarityDataTraits = rarityData[traitType];
-    rarityDataTraits.forEach((rarityDataTrait) => {
-      if (rarityDataTrait.trait == value) {
-        // keep track of occurrences
-        rarityDataTrait.occurrence++;
-      }
+  // if the total_score is true then add the total_score to the output
+  if (rarityConfigurations.total_score) {
+    output.forEach((nft) => {
+      nft.total_score =
+        Math.round(nft.attributes.reduce((a, b) => a + b.score, 0) * 100) / 100;
     });
-  });
-});
+  }
```

Comment:
The ordering of these two also need to be switched since you're getting rank from total score.